read-fonts/src/variations.rs

```diff
@@ -0,0 +1,200 @@
+//! OpenType font variations common tables.
+
+include!("../generated/generated_variations.rs");
+
+/// Outer and inner indices for reading from an [ItemVariationStore].
+#[derive(Copy, Clone, Debug)]
+pub struct DeltaSetIndex {
+    /// Outer delta set index.
+    pub outer: u16,
+    /// Inner delta set index.
+    pub inner: u16,
+}
+
+impl<'a> DeltaSetIndexMap<'a> {
+    /// Returns the delta set index for the specified value.
+    pub fn get(&self, index: u32) -> Result<DeltaSetIndex, ReadError> {
+        let (entry_format, data) = match self {
+            Self::Format0(fmt) => (fmt.entry_format() as u32, fmt.map_data()),
+            Self::Format1(fmt) => (fmt.entry_format() as u32, fmt.map_data()),
+        };
+        let entry_size = ((entry_format & EntryFormat::MAP_ENTRY_SIZE_MASK.bits() as u32) >> 4) + 1;
+        let data = FontData::new(data);
+        let offset = index as usize * entry_size as usize;
+        let entry = match entry_size {
+            1 => data.read_at::<u8>(offset)? as u32,
+            2 => data.read_at::<u16>(offset)? as u32,
+            3 => data.read_at::<Uint24>(offset)?.into(),
+            4 => data.read_at::<u32>(offset)?,
+            _ => {
+                return Err(ReadError::MalformedData(
+                    "invalid entry size in DeltaSetIndexMap",
+                ))
+            }
+        };
+        let bit_count = (entry_format & 0xF) + 1;
+        Ok(DeltaSetIndex {
+            outer: (entry >> bit_count) as u16,
+            inner: (entry & ((1 << bit_count) - 1)) as u16,
+        })
+    }
+}
+
+fn map_data_size(entry_format: u8, map_count: u32) -> usize {
+    (((entry_format & 0x30) >> 4) + 1) as usize * map_count as usize
+}
+
+impl<'a> ItemVariationStore<'a> {
+    /// Computes the delta value for the specified index and set of normalized
+    /// variation coordinates.
+    pub fn compute_delta(
+        &self,
+        index: DeltaSetIndex,
+        coords: &[F2Dot14],
+    ) -> Result<Fixed, ReadError> {
+        let data = match self
+            .item_variation_datas()
+            .nth(index.outer as usize)
+            .flatten()
+        {
+            Some(data) => data?,
+            None => return Ok(Fixed::default()),
+        };
+        let regions = self.variation_region_list()?.variation_regions();
+        let region_indices = data.region_indexes();
+        let mut delta = Fixed::ZERO;
+        for (i, region_delta) in data.delta_set(index.inner).enumerate() {
+            let region_index = region_indices
+                .get(i)
+                .ok_or(ReadError::MalformedData(
+                    "invalid delta sets in ItemVariationStore",
```

Comment:
It would be nice to include specifics such as the index and the value of i. In Python or Java we'd stick it in the message but perhaps in Rust it would be more normal to have actual fields on MalformedData?

Reply:
Same comment for "invalid entry size in DeltaSetIndexMap" above

Reply:
We can add another variant to ReadError (I added one for missing metric tags in MVAR), but I'm not sure how far we want to go down that path

Reply:
A MalformedDataAtIndex error with a field for the offending index seems like it would go quite far. Make the index an enum with variants for at least naked u32 and DeltaSetIndex and you've covered a *lot* of cases.